### PR TITLE
feat: two-stage protocol service initialization & inbound dispatcher package

### PR DESCRIPTION
**Pull request by Moopli**
Signed-off-by: Filip Burlacu <filip.burlacu@securekey.com>

### CODE REVIEW

**Comment by Moopli** [comment]
Marking as ready, only needs test coverage.

File: pkg/didcomm/dispatcher/inbound/inbound_message_handler.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	}
+
+	isV2, err := service.IsDIDCommV2(&msg)
+	if err != nil {
+		return err
+	}
+
+	// if msg is a didcomm v2 message, do additional handling
+	if isV2 {
+		logger.Debugf("didcomm v2 message handling before service...")
```
Comment (Baha-sk) [inline pkg/didcomm/dispatcher/inbound/inbound_message_handler.go:101]: add a TODO here?

File: pkg/didcomm/dispatcher/inbound/inbound_message_handler.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+			return err
+		}
+	}
+
+	// in case of no services are registered for given message type,
+	// find generic inbound services registered for given message header
+	for _, svc := range handler.msgSvcProvider.Services() {
+		h := struct {
+			Purpose []string `json:"~purpose"`
```
Comment (Baha-sk) [inline pkg/didcomm/dispatcher/inbound/inbound_message_handler.go:129]: is this a V1 format?

File: pkg/didcomm/dispatcher/inbound/inbound_message_handler.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+			return handler.tryToHandle(svc, msg, service.NewDIDCommContext(myDID, theirDID, nil))
+		}
+	}
+
+	return fmt.Errorf("no message handlers found for the message type: %s", msg.Type())
+}
+
+//nolint:nestif,gocognit,funlen,gocyclo
+func (handler *MessageHandler) getDIDs(envelope *transport.Envelope) (string, string, error) {
```
Comment (Baha-sk) [inline pkg/didcomm/dispatcher/inbound/inbound_message_handler.go:151]: just to confirm, this was extract as-is from the original inboundhandler, correct?

File: pkg/didcomm/protocol/mediator/service.go [hunk 4]
```diff
@@ -149,55 +149,77 @@ type Service struct {
...
 	mp, err := prov.Service(messagepickup.MessagePickup)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	messagePickupSvc, ok := mp.(messagepickup.ProtocolService)
 	if !ok {
-		return nil, errors.New("cast service to message pickup service failed")
+		return errors.New("cast service to message pickup service failed")
```
Comment (Baha-sk) [inline pkg/didcomm/protocol/mediator/service.go:201]: why not add Type check in the error here as  was done for the providers above?

File: pkg/didcomm/protocol/outofband/service.go [hunk 5]
```diff
@@ -156,59 +157,80 @@ type Provider interface {
...
 	svc, err := p.Service(didexchange.DIDExchange)
 	if err != nil {
-		return nil, fmt.Errorf("failed to initialize outofband service : %w", err)
+		return fmt.Errorf("failed to initialize outofband service : %w", err)
 	}
 
 	didSvc, ok := svc.(didExchSvc)
 	if !ok {
-		return nil, errors.New("failed to cast the didexchange service to satisfy our dependency")
+		return errors.New("failed to cast the didexchange service to satisfy our dependency")
```
Comment (Baha-sk) [inline pkg/didcomm/protocol/outofband/service.go:188]: same, simply say something like: oob depends on service didExchangeService, got (%T) instead

File: pkg/didcomm/dispatcher/inbound/inbound_message_handler.go (hunk 3 above)
Comment (Moopli) [reply pkg/didcomm/dispatcher/inbound/inbound_message_handler.go:151]: Almost completely as-is, I'll note the only change

File: pkg/didcomm/dispatcher/inbound/inbound_message_handler.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+	return myDID, theirDID, backoff.Retry(func() error {
+		var notFound bool
+
+		// nolint: gocritic
+		if id := strings.Index(string(envelope.ToKey), kaIdentifier); id > 0 &&
+			strings.Index(string(envelope.ToKey), "\"kid\":\"did:") > 0 {
+			myDID, err = pubKeyToDID(envelope.ToKey)
+			if err != nil {
+				return backoff.Permanent(fmt.Errorf("getMyDID: envelope.ToKey as myDID: %w", err))
```
Comment (Moopli) [inline pkg/didcomm/dispatcher/inbound/inbound_message_handler.go:166]: @Baha-sk the only change is that these errors are now wrapped in `backoff.Permanent()`. If pubKeyToDID fails, then it's never going to be able to succeed in a retry, so wrapping it in `backoff.Permanent` keeps it from retrying a case that can only fail over and over.

**Review by Baha-sk — APPROVED** [review]

File: pkg/didcomm/dispatcher/inbound/inbound_message_handler.go (hunk 2 above)
Comment (Moopli) [reply pkg/didcomm/dispatcher/inbound/inbound_message_handler.go:129]: Yes - I'm not sure what will happen with the message service and purpose decorator when/if it gets ported to didcomm/v2, but this part is v1 only.

I can put it inside an `if`.

Comment (Baha-sk) [reply pkg/didcomm/dispatcher/inbound/inbound_message_handler.go:129]: yes, it would be better

**Review by fqutishat — APPROVED** [review]

File: pkg/didcomm/protocol/mediator/service.go (hunk 4 above)
Comment (Moopli) [reply pkg/didcomm/protocol/mediator/service.go:201]: minor issue, baha agreed there's no need

File: pkg/didcomm/protocol/outofband/service.go (hunk 5 above)
Comment (Moopli) [reply pkg/didcomm/protocol/outofband/service.go:188]: minor issue, baha agreed there's no need

File: pkg/didcomm/dispatcher/inbound/inbound_message_handler.go (hunk 2 above)
Comment (Moopli) [reply pkg/didcomm/dispatcher/inbound/inbound_message_handler.go]: done

**Pull request merged**